# Introduction to scikit-learn (sklearn)
Demonstrate how scikit-learn works

0. An end-to-end scikit-learn workflow
1. Getting the data ready
2. Choose the right estimator/algorithm for our problems
3. Fit the model/algorithm and use it to make predictions on our data
4. Evaluating a model
5. Improve a model
6. Save and load a trained model
7. Putting it all together!

## 0. An end-to-end scikit-learn workflow

## 1. Getting the data ready

In [11]:
import pandas as pd
import numpy as np
heart_disease = pd.read_csv("../heart-disease.csv")
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  ca  thal  target
0     63    1   3       145   233    1        0      150      0      2.3      0   0     1       1
1     37    1   2       130   250    0        1      187      0      3.5      0   0     2       1
2     41    0   1       130   204    0        0      172      0      1.4      2   0     2       1
3     56    1   1       120   236    0        1      178      0      0.8      2   0     2       1
4     57    0   0       120   354    0        1      163      1      0.6      2   0     2       1
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...    ...  ..   ...     ...
298   57    0   0       140   241    0        1      123      1      0.2      1   0     3       0
299   45    1   3       110   264    0        1      132      0      1.2      1   0     3       0
300   68    1   0       144   193    1        1      141      0      3.4      1   2     3       0
301   57    1   0       130   131    0        1      115      1      1.2      1   1     3       0
302   57    0   1       130   236    0        0      174      0      0.0      1   1     2       0

[303 rows x 14 columns]

In [2]:
# Create x (features matrix)
x = heart_disease.drop("target", axis=1)

# Create y (labels)
y = heart_disease["target"]

## 2. Choose the right model and hyperparameters

In [6]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()

# Keep the default hyperparameters
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

## 3. Fit the model to the training data

In [7]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [10]:
clf.fit(x_train, y_train);

In [13]:
y_preds = clf.predict(x_test)

In [14]:
y_preds

array([0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0], dtype=int64)

In [15]:
y_test

266    0
146    1
237    0
164    1
159    1
      ..
147    1
132    1
176    0
276    0
219    0
Name: target, Length: 61, dtype: int64

## 4. Evaluate the model on the training data and test data

In [16]:
clf.score(x_train, y_train)

1.0

### The model scored 1 out of 1 on the test data, which is very good. 

In [17]:
clf.score(x_test, y_test)

0.8360655737704918

### The model scored correct on 83% of the test data (data that the model has not seen before)

In [19]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.79      0.85      0.81        26
           1       0.88      0.83      0.85        35

    accuracy                           0.84        61
   macro avg       0.83      0.84      0.83        61
weighted avg       0.84      0.84      0.84        61



In [20]:
confusion_matrix(y_test, y_preds)

array([[22,  4],
       [ 6, 29]], dtype=int64)

In [21]:
accuracy_score(y_test, y_preds)

0.8360655737704918

## 5. Improve the model

In [24]:
# Try differente amount of n_estimators
np.random.seed(42)
for i in range(10, 100, 10): 
    print(f"Trying model with {i} estimators...")
    clf = RandomForestClassifier(n_estimators=i).fit(x_train, y_train)
    print(f"Model accuracy on test set: {clf.score(x_test, y_test) * 100:.2f}%")
    print("")

Trying model with 10 estimators...
Model accuracy on test set: 83.61%

Trying model with 20 estimators...
Model accuracy on test set: 81.97%

Trying model with 30 estimators...
Model accuracy on test set: 83.61%

Trying model with 40 estimators...
Model accuracy on test set: 81.97%

Trying model with 50 estimators...
Model accuracy on test set: 83.61%

Trying model with 60 estimators...
Model accuracy on test set: 81.97%

Trying model with 70 estimators...
Model accuracy on test set: 80.33%

Trying model with 80 estimators...
Model accuracy on test set: 83.61%

Trying model with 90 estimators...
Model accuracy on test set: 81.97%



### Choosing more estimators does not improve the model from the original. 83.61% is the best we can get using 10 estimators. 

In [28]:
clf = RandomForestClassifier(n_estimators=10).fit(x_train, y_train)

## 6. Save the model and load it

In [29]:
import pickle

pickle.dump(clf, open("random_forest_model_1.pkl", "wb")) # wb = write binary

In [30]:
loaded_model = pickle.load(open("random_forest_model_1.pkl", "rb")) # rb = read binary
loaded_model.score(x_test, y_test)

0.8360655737704918